ext/liquid_c/vm.c

```diff
@@ -11,53 +12,394 @@ ID id_ivar_last_capture_length;
+            write_fixnum(output, obj);
+            break;
+        case T_ARRAY:
+            for (long i = 0; i < RARRAY_LEN(obj); i++)
+            {
+                VALUE item = RARRAY_AREF(obj, i);
+
+                if (RB_UNLIKELY(RB_TYPE_P(item, T_ARRAY))) {
+                    // Normally liquid arrays are flat, but for safety and simplicity we
+                    // leverage ruby's join that detects and raises on a recursion loop
+                    rb_str_buf_append(output, rb_ary_join(item, Qnil));
+                } else {
+                    write_obj(output, item);
+                }
+            }
+            break;
+        case T_NIL:
+            break;
+        default:
+            obj = rb_funcall(obj, id_to_s, 0);
+            rb_str_append(output, obj);
+            break;
     }
+}
+
+static inline void vm_stack_push(vm_t *vm, VALUE value)
+{
+    VALUE *stack_ptr = (VALUE *)vm->stack.data_end;
+    *stack_ptr++ = value;
+    vm->stack.data_end = (uint8_t *)stack_ptr;
+}
 
-    const size_t *const_ptr = (const size_t *)body->constants.data;
-    const uint8_t *ip = body->instructions.data;
-    VALUE interrupts = rb_ivar_get(context, id_ivar_interrupts);
-    Check_Type(interrupts, T_ARRAY);
+static inline VALUE vm_stack_pop(vm_t *vm)
+{
+    VALUE *stack_ptr = (VALUE *)vm->stack.data_end;
+    stack_ptr--;
+    vm->stack.data_end = (uint8_t *)stack_ptr;
+    return *stack_ptr;
+}
+
+static inline VALUE *vm_stack_pop_n_use_in_place(vm_t *vm, size_t n)
+{
+    VALUE *stack_ptr = (VALUE *)vm->stack.data_end;
+    stack_ptr -= n;
+    vm->stack.data_end = (uint8_t *)stack_ptr;
+    return stack_ptr;
+}
+
+static inline void vm_stack_reserve_for_write(vm_t *vm, size_t num_values)
+{
+    c_buffer_reserve_for_write(&vm->stack, num_values * sizeof(VALUE));
+}
+
+static VALUE vm_invoke_filter(vm_t *vm, VALUE filter_name, int num_args, VALUE *args)
+{
+    bool not_invokable = rb_hash_lookup(vm->filter_methods, filter_name) != Qtrue;
+    if (RB_UNLIKELY(not_invokable)) {
+        if (vm->strict_filters) {
+            VALUE error_class = rb_const_get(mLiquid, rb_intern("UndefinedFilter"));
+            rb_raise(error_class, "undefined filter %"PRIsVALUE, rb_sym2str(filter_name));
+        }
+        return args[0];
+    }
+
+    VALUE result = rb_funcallv(vm->strainer, RB_SYM2ID(filter_name), num_args, args);
+    return rb_funcall(result, id_to_liquid, 0);
+}
+
+typedef struct vm_render_until_error_args {
+    vm_t *vm;
+    const uint8_t *ip; // use for initial address and to save an address for rescuing
+    const size_t *const_ptr;
+    unsigned int node_line_number;
+    VALUE context;
+    VALUE output;
+    bool is_captured;
+    long render_length_limit;
+} vm_render_until_error_args_t;
+
+static inline void increment_write_score(vm_render_until_error_args_t *args)
+{
+    if (RB_UNLIKELY(args->is_captured)) {
+        rb_funcall(args->vm->resource_limits, id_increment_write_score, 1, args->output);
```

Comment:
I had actually marked this as an unlikely code path using `RB_UNLIKELY` since this is only when writing to the capture buffer.  When writing directly to the output buffer, we will have the much faster check on the following line which is just a bounds check.  https://github.com/Shopify/liquid/pull/1285 provides some context on why we can't just do the same optimization for capture buffers.

There are opportunities for optimizing these checks, but I think it is outside the scope of this PR.  This PR just moves it to an inline function so that we are only doing the checks after write instructions rather than after all instructions.

Reply:
Ah I see. I read the code wrong, missed the bounds check. 👍